rest-server/src/models/user.js

```diff
@@ -133,6 +134,93 @@ const remove = (username, callback) => {
   }
 };
 
+const updateUserVc = (username, virtualClusters, callback) => {
+  if (typeof username === 'undefined') {
+    callback(new Error('user does not exist'), false);
+  } else {
+    db.get(etcdConfig.userPath(username), null, (errMsg, res) => {
+      if (errMsg) {
+        logger.warn('user %s not exists', etcdConfig.userPath(username));
+        callback(errMsg, false);
+      } else {
+        VirtualCluster.prototype.getVcList((vcList, err) => {
+          if (err) {
+            logger.warn('get virtual cluster list error\n%s', err.stack);
+          } else if (vcList === undefined) {
+            logger.warn('list virtual clusters error, no virtual cluster found');
+          } else {
+            let updateVcList = (res.get(etcdConfig.userAdminPath(username)) === 'true') ? Object.keys(vcList) : virtualClusters.trim().split(',').filter((updateVc) => (updateVc !== ''));
+            let addUserWithInvalidVc = false;
+            for (let item of updateVcList) {
+              if (!vcList.hasOwnProperty(item)) {
+                if (!res.has(etcdConfig.userVirtualClusterPath(username))) {
+                  updateVcList.length = 0;
+                  addUserWithInvalidVc = true;
+                  break;
+                } else {
+                  return callback(new Error('InvalidVirtualCluster'), false);
+                }
+              }
+            }
+            if (!updateVcList.includes('default')) { // always has 'default' queue
+              updateVcList.push('default');
+            }
+            updateVcList.sort();
+            db.set(etcdConfig.userVirtualClusterPath(username), updateVcList.toString(), null, (errMsg, res) => {
+              if (errMsg) {
+                logger.warn('update %s virtual cluster: %s failed, error message:%s', etcdConfig.userVirtualClusterPath(username), errMsg);
+                callback(errMsg, false);
+              } else {
+                if (addUserWithInvalidVc) {
```

Comment:
why use closure here and why put this in db.set callback?

Reply:
When add new user if input virtual cluster is invalid we also need to add this user to default group, and we also should let user know the input is invalid.

Reply:
@asakuri: I discussed with Can yesterday. Her philosophy is to add user with best efforts. Do you have some examples of pros and cons of rejecting the whole request? 

And @can: I suggest you include the VCs that are not valid in the http response, we should not do anything in the background out of admin’s expectation.